core/server/master/src/main/java/alluxio/master/file/meta/TrieNode.java

```diff
@@ -0,0 +1,252 @@
+   * @param key the target TrieNode's key
+   */
+  public void removeChild(T key) {
+    mChildren.remove(key);
+  }
+
+  /**
+   * add child to the current TrieNode.
+   * @param key the target key
+   * @param value the target value(TrieNode)
+   */
+  public void addChild(T key, TrieNode<T> value) {
+    mChildren.put(key, value);
+  }
+
+  /**
+   * get the child TrieNode by given key.
+   * @param key the given key to get the corresponding child
+   * @return the corresponding child TrieNode
+   */
+  public TrieNode<T> child(T key) {
+    return mChildren.get(key);
+  }
+
+  /**
+   * acquire all descendant TrieNodes.
+   *
+   * @param isNodeMustTerminal true if the descendant node must also be a terminal node
+   * @param isContainSelf true if the results can contain itself
+   * @param terminateAfterAdd true if the search terminates instantly after finding one qualified
+   * @return all the children TrieNodes
+   */
+  public List<TrieNode<T>> descendants(boolean isNodeMustTerminal, boolean isContainSelf,
+                                       boolean terminateAfterAdd) {
+    List<TrieNode<T>> childrenNodes = new ArrayList<>();
+
+    // For now, we use BFS to acquire all nested TrieNodes underneath the current TrieNode.
+    Queue<TrieNode<T>> queue = new LinkedList<>();
+    queue.add(this);
+
+    while (!queue.isEmpty()) {
+      TrieNode<T> front = queue.poll();
+      // checks if the front of the queue can pass both the terminal check, and the check on
+      // whether regarding itself as a descendants.
+      if (front.checkNodeTerminal(isNodeMustTerminal) && (isContainSelf || front != this)) {
+        childrenNodes.add(front);
+        if (terminateAfterAdd) {
+          break;
+        }
+      }
+      // adds all children of front into the queue.
+      for (Map.Entry<T, TrieNode<T>> entry : front.mChildren.entrySet()) {
+        TrieNode<T> value = entry.getValue();
+        queue.add(value);
+      }
+    }
+    return childrenNodes;
+  }
+
+  /**
+   * Checks if current TrieNode contains the certain type of TrieNode.
+   *
+   * @param isContainSelf true if the results may contain current TrieNode
+   * @return true if current TrieNode has children that match the given filters
+   */
+  public boolean hasNestedTerminalTrieNodes(boolean isContainSelf) {
+    List<TrieNode<T>> descendants = descendants(true, isContainSelf, true);
+    return descendants.size() > 0;
+  }
+
+  /**
+   * Removes the given nodes from current TrieNode. The given values must correspond to a
+   * terminal TrieNode.
+   *
+   * @param values inodes of the path to be removed
+   * @return the removed terminal node if the inodes are removed successfully, else return null
+   */
+  public TrieNode<T> remove(List<T> values) {
+    // parents store several <TrieNode, T> pairs, each pair contains the parent TrieNode and the
+    // value of its child along the given values.
+    Stack<Pair<TrieNode<T>, T>> parents = new Stack<>();
```

Comment:
What is this method doing? Does it remove `List<T> values` as a path in the Trie, or as multiple TrieNode instances which are all in `this.mChildren`?